crates/uv-resolver/src/lock/export/cyclonedx_json.rs

```diff
@@ -0,0 +1,417 @@
+        tool::{Tool, Tools},
+    },
+    prelude::{Bom, Component, Components, NormalizedString},
+};
+use itertools::Itertools;
+use percent_encoding::{AsciiSet, CONTROLS, percent_encode};
+use rustc_hash::FxHashSet;
+
+use uv_configuration::{
+    DependencyGroupsWithDefaults, ExtrasSpecificationWithDefaults, InstallOptions,
+};
+use uv_fs::PortablePath;
+use uv_normalize::PackageName;
+use uv_pep508::MarkerTree;
+use uv_preview::{Preview, PreviewFeatures};
+use uv_warnings::warn_user;
+
+use crate::lock::export::{ExportableRequirement, ExportableRequirements};
+use crate::lock::{LockErrorKind, Package, PackageId, Source};
+use crate::{Installable, LockError};
+
+/// Character set for percent-encoding PURL components, copied from packageurl.rs (<https://github.com/scm-rs/packageurl.rs/blob/a725aa0ab332934c350641508017eb09ddfa0813/src/purl.rs#L18>).
+const PURL_ENCODE_SET: &AsciiSet = &CONTROLS
+    .add(b' ')
+    .add(b'"')
+    .add(b'#')
+    .add(b'%')
+    .add(b'<')
+    .add(b'>')
+    .add(b'`')
+    .add(b'?')
+    .add(b'{')
+    .add(b'}')
+    .add(b';')
+    .add(b'=')
+    .add(b'+')
+    .add(b'@')
+    .add(b'\\')
+    .add(b'[')
+    .add(b']')
+    .add(b'^')
+    .add(b'|');
+
+/// Creates `CycloneDX` components, registering them in a `HashMap` so that they can be retrieved by `PackageId`.
+/// Also ensures uniqueness when generating bom-refs by using a numeric prefix which is incremented for each component.
+#[derive(Default)]
+struct ComponentBuilder<'a> {
+    id_counter: usize, // Used as prefix in bom-ref generation, to ensure uniqueness
+    package_to_component_map: HashMap<&'a PackageId, Component>,
+}
+
+impl<'a> ComponentBuilder<'a> {
+    /// Creates a bom-ref string in the format "{id}-{package_name}@{version}" or "{id}-{package_name}" if no version is provided.
+    fn create_bom_ref(&mut self, name: &str, version: Option<&str>) -> String {
+        self.id_counter += 1;
+        let id = self.id_counter;
+        if let Some(version) = version {
+            format!("{name}-{id}@{version}")
+        } else {
+            format!("{name}-{id}")
+        }
+    }
+
+    /// Extract version string from a package.
+    fn get_version_string(package: &Package) -> Option<String> {
+        package
+            .id
+            .version
+            .as_ref()
+            .map(std::string::ToString::to_string)
+    }
+
+    /// Extract package name string from a package.
+    fn get_package_name(package: &Package) -> &str {
+        package.id.name.as_str()
+    }
+
+    /// Generate a Package URL (purl) from a package. Returns `None` for local sources.
+    fn create_purl(package: &Package) -> Option<String> {
+        let name = percent_encode(Self::get_package_name(package).as_bytes(), PURL_ENCODE_SET);
+
+        let version = Self::get_version_string(package).map_or_else(String::new, |v| {
+            format!("@{}", percent_encode(v.as_bytes(), PURL_ENCODE_SET))
+        });
```

Comment:
This can use `unwrap_or_default()`, or some form of matching, I find this one a bit hard to follow.